classes/file_search.php

```diff
@@ -344,6 +347,45 @@ public static function grep_content($csv, $filecontents, $criteria, $stream): in
         return $matchcount;
     }
 
+    /**
+     * Handles processing of a grep search by splitting files into smaller chunks when required.
+     *
+     * @param array $csv Some columns to be output in the csv file.
+     * @param \stored_file|\ZipArchive $file The file or zip archive.
+     * @param object $criteria The regular expression to be matched.
+     * @param resource $stream The stream handle for the output file.
+     * @param array $zipstat information about the file inside a zip full.
+     * @return int $matchcount The number of matches found.
+     */
+    public static function grep_processor($csv, $file, $criteria, $stream, $zipstat = []) {
+        $storedfile = $file instanceof \stored_file;
+
+        // For files smaller than chunk size, just open them directly.
+        $filesize = $storedfile ? $file->get_filesize() : $zipstat['size'];
+        if ($filesize < self::CHUNK_SIZE) {
+            $content = $storedfile ? $file->get_content() : $file->getFromIndex($zipstat['index']);
+            return self::grep_content($csv, $content, $criteria, $stream);
+        }
+
+        // For large files, use a file handle stream instead.
+        $matchcount = 0;
+        $handle = $storedfile ? $file->get_content_file_handle() : $file->getStream($file->getNameIndex($zipstat['index']));
+        if (empty($handle)) {
+            return $matchcount;
+        }
+
+        $prev = '';
+        while (!feof($handle)) {
+            // Prepend the last few characters to the next iteration so no border cases are missed.
+            $chunk = $prev . fread($handle, self::CHUNK_SIZE);
```

Comment:
Does the performance get much worse with a small chunk size? Wondering what the impact is if we used fgets instead of fread which brings us one step closer to implementing columns for line and column numbers  https://github.com/catalyst/moodle-tool_advancedreplace/issues/58